File: mission17-backend/ai-perf-test.py
import requests
import threading
import time
import os

URL = 'http://localhost:5000/predict'
PENDING_COUNT = 5

# Find a test image
UPLOADS_DIR = 'uploads'
test_image_path = None

if os.path.exists(UPLOADS_DIR):
    for f in os.listdir(UPLOADS_DIR):
        if f.endswith('.jpg'):
            test_image_path = os.path.join(UPLOADS_DIR, f)
            break

if not test_image_path:
    # Create a dummy image file with junk data if no real image exists
    test_image_path = 'dummy_test.jpg'
    with open(test_image_path, 'wb') as f:
        f.write(b'\xFF\xD8\xFF\xE0\x00\x10JFIF\x00\x01\x01\x01\x00H\x00H\x00\x00\xFF\xDB\x00C\x00')

def make_request(request_id, results):
    try:
        with open(test_image_path, 'rb') as img_file:
            files = {'file': ('proof.jpg', img_file, 'image/jpeg')}
            
            # Start timer for this specific request
            start_req = time.time()
            response = requests.post(URL, files=files)
            end_req = time.time()
            
            if response.status_code == 200:
                results['success'] += 1
                # print(f"Request {request_id} SUCCESS in {end_req - start_req:.2f}s")
            else:
                results['failed'] += 1
                # print(f"Request {request_id} FAILED ({response.status_code}) in {end_req - start_req:.2f}s")
                
    except Exception as e:
        results['failed'] += 1
        # print(f"Request {request_id} ERROR: {str(e)}")

print("\n🤖 --- AI CONCURRENCY BENCHMARK ---")

# =========================================================================
# TEST 1: The "Bad" Way (Parallel requests simulating Promise.all)
# =========================================================================
print(f"\n[TEST 1: THE BOTTLENECK] Hitting the AI Server with {PENDING_COUNT} parallel requests (Simulating Promise.all)...")

results_parallel = {'success': 0, 'failed': 0}
threads = []

start_parallel = time.time()

# Fire all requests at the exact same time
for i in range(PENDING_COUNT):
    t = threading.Thread(target=make_request, args=(i, results_parallel))
    threads.append(t)
    t.start()

# Wait for all to finish
for t in threads:
    t.join()

end_parallel = time.time()

print(f"   Time taken: {(end_parallel - start_parallel) * 1000:.2f} ms")
print(f"   Successful analyses: {results_parallel['success']}")
print(f"   Failed inferences: {results_parallel['failed']}")

if results_parallel['failed'] > 0:
    print("   ⚠️ Look! The AI server dropped requests/crashed under parallel load!")

time.sleep(2) # Let the server catch its breath

# =========================================================================
# TEST 2: The "Good" Way (Sequential requests simulating the fix)
# =========================================================================
print(f"\n[TEST 2: THE FIX] Hitting the AI Server with {PENDING_COUNT} sequential requests...")

results_sequential = {'success': 0, 'failed': 0}
start_sequential = time.time()

# Fire requests one after the other
for i in range(PENDING_COUNT):
    make_request(i, results_sequential)

end_sequential = time.time()

print(f"   Time taken: {(end_sequential - start_sequential) * 1000:.2f} ms")
print(f"   Successful analyses: {results_sequential['success']}")
print(f"   Failed inferences: {results_sequential['failed']}")

print("\n✅ Benchmark complete.")

File: mission17-backend/ai-perf-test.js
import fs from 'fs';
import path from 'path';
import 'dotenv/config';

async function buildAIFormData(imageUri) {
  const imageBuffer = fs.readFileSync('fake_image.jpg');
  const formData = new FormData();
  const blob = new Blob([imageBuffer], { type: 'image/jpeg' });
  formData.append('file', blob, 'proof.jpg');
  return formData;
}

// Mimics the exact fetch call inside your submissions.js
async function callAIServer(imageUri) {
  const formData = await buildAIFormData(imageUri);
  const aiResponse = await fetch(process.env.AI_SERVER_URL || 'http://localhost:5000/predict', {
    method: 'POST',
    body: formData,
  });

  if (!aiResponse.ok) {
    const errText = await aiResponse.text();
    throw new Error(`AI Server Error (${aiResponse.status}): ${errText}`);
  }
  return aiResponse.json();
}

async function runTest() {
  console.log('🤖 --- AI CONCURRENCY BENCHMARK ---\n');
  const PENDING_COUNT = 5; // Simulate 5 submissions waiting for AI analysis
  
  const dummySubmissions = Array.from({ length: PENDING_COUNT }).map((_, i) => ({
    _id: `submission-${i}`,
    imageUri: 'dummy_test.jpg'
  }));

  // =========================================================================
  // TEST 1: The "Bad" Way (Parallel via Promise.all) - currently in code
  // =========================================================================
  console.log(`[TEST 1: THE BOTTLENECK] Hitting the AI Server with ${PENDING_COUNT} parallel requests using Promise.all...`);
  const startParallel = performance.now();
  let parallelSuccessCount = 0;
  let parallelFailCount = 0;
  
  try {
    await Promise.all(
        dummySubmissions.map(async sub => {
          try {
            await callAIServer(sub.imageUri);
            parallelSuccessCount++;
          } catch (e) {
            parallelFailCount++;
          }
        })
    );
  } catch (e) {
      console.log('Fatal Error in Promise.all loop');
  }

  const endParallel = performance.now();
  const timeParallel = (endParallel - startParallel).toFixed(2);
  
  console.log(`   Time taken: ${timeParallel} ms`);
  console.log(`   Successful analyses: ${parallelSuccessCount}`);
  console.log(`   Failed inferences: ${parallelFailCount}`);
  if (parallelFailCount > 0) {
      console.log('   ⚠️ Look! The AI server dropped requests/crashed under parallel load!');
  }


  // =========================================================================
  // TEST 2: The "Good" Way (Sequential Execution) - proposed fix
  // =========================================================================
  console.log(`\n[TEST 2: THE FIX] Hitting the AI Server with ${PENDING_COUNT} sequential requests...`);
  const startSequential = performance.now();
  let sequentialSuccessCount = 0;
  let sequentialFailCount = 0;

  for (const sub of dummySubmissions) {
      try {
          await callAIServer(sub.imageUri);
          sequentialSuccessCount++;
      } catch (e) {
          console.error(e.message);
          sequentialFailCount++;
      }
  }

  const endSequential = performance.now();
  const timeSequential = (endSequential - startSequential).toFixed(2);
  
  console.log(`   Time taken: ${timeSequential} ms`);
  console.log(`   Successful analyses: ${sequentialSuccessCount}`);
  console.log(`   Failed inferences: ${sequentialFailCount}`);
  
  console.log('\n✅ Benchmark complete.');
}

runTest();
